.ci/windows-steps.yaml

```diff
@@ -12,8 +12,10 @@ steps:
     nuget install boost -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install boost_random-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install boost_math_c99-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
-    nuget install OpenBLAS -o $(Agent.ToolsDirectory)
     nuget install unofficial-flayan-cereal -o $(Agent.ToolsDirectory)
+    nuget install ensmallen -o $(Agent.ToolsDirectory) -Version 2.17.0
+    dir $(Agent.ToolsDirectory)
+    tree $(Agent.ToolsDirectory)
 
     mkdir -p $(Agent.ToolsDirectory)/boost_libs
     cp $(Agent.ToolsDirectory)/boost_math_c99-vc140.1.60.0.0/lib/native/address-model-64/lib/*.* $(Agent.ToolsDirectory)/boost_libs
```

Comment:
I know this was not part of your PR, but I'm pretty sure if you change `tar -xvzf` to `tar -xvf` below, that should download the Armadillo source correctly, and then we can see if using `tmp/include/` as the include directory succeeds. :+1:

Reply:
Interesting observation to add to this pull request:
In addition to this pull request, I have added `message(STATUS "ARMADILLO LIBRARY" ${ARMADILLO_LIBRARIES})` in this pull request #3040 
As you can see, it did not print the armadillo library location, but instead it did print the OpenBLAS location

https://dev.azure.com/mlpack/mlpack/_build/results?buildId=7268&view=logs&j=934eac7e-378e-5e05-ed87-f05e34fa9fdc&t=9715bafe-ea9f-54fa-e80f-ab012eb34e65&l=34

I think there is a small bug here in the armadillo deployment on Windows that is passing everywhere, but it detected somehow with this pull request.

Reply:
Thanks for linking to that.  I looked through the CMake code and I am not entirely sure what is going on, but I think I at least have a partial solution.  Armadillo has two linking modes, controlled by whether the macro `ARMA_USE_WRAPPER` is set or not.  On Windows, if `ARMA_USE_WRAPPER` is set, we must link against Armadillo *and* OpenBLAS; if it is not set, we only need to link against OpenBLAS.

By default, Armadillo ships with a file `armadillo_bits/config.hpp` that leaves `ARMA_USE_WRAPPER` *unset*, in the `include/` directory.  When we run CMake in the Armadillo sources, this generates a directory `tmp/`, and inside that directory there will be a `tmp/include/armadillo_bits/config.hpp` that *will* have `ARMA_USE_WRAPPER` set.

However, I can see in our choice of `ARMADILLO_INCLUDE_DIR` below that we are *not* using the `tmp/` directory, so the mlpack file `CMake/FindArmadillo.cmake` should detect that `ARMA_USE_WRAPPER` is *not* set.  Then, this code runs:

```
if(_ARMA_USE_WRAPPER)
  # UNIX paths are standard, no need to write.
  find_library(ARMADILLO_LIBRARY
    NAMES armadillo
    PATHS "$ENV{ProgramFiles}/Armadillo/lib"  "$ENV{ProgramFiles}/Armadillo/lib64" "$ENV{ProgramFiles}/Armadillo"
    )
    set(_ARMA_REQUIRED_VARS ARMADILLO_LIBRARY ARMADILLO_INCLUDE_DIR VERSION_VAR ARMADILLO_VERSION_STRING)
else()
  # don't link to armadillo in this case
  set(ARMADILLO_LIBRARY "")
endif()
```

So, in our case, that should set `ARMADILLO_LIBRARY` to the empty string, which is why it does not appear in `ARMADILLO_LIBRARIES`.  This code runs at the end of `CMake/FindArmadillo.cmake`:

```
if (ARMADILLO_FOUND)
  set(ARMADILLO_INCLUDE_DIRS ${ARMADILLO_INCLUDE_DIR})
  set(ARMADILLO_LIBRARIES ${ARMADILLO_LIBRARY} ${_ARMA_SUPPORT_LIBRARIES})
endif ()
```

So, we can see based on the CMake output in the log you linked to that we are configuring mlpack as if Armadillo has `ARMA_USE_WRAPPER` *unset*.

But, I am confused, because then when linking we get all kinds of errors like this:

```
2021-09-08T22:50:13.7010984Z   mlpack.lib(lcc.obj) : error LNK2001: unresolved external symbol wrapper2_dgemm_ [D:\a\1\s\build\src\mlpack\tests\mlpack_test.vcxproj]
```

That's confusing because `_wrapper2_dgemm` shouldn't ever appear unless `ARMA_USE_WRAPPER` is set!

So, in conclusion, I'm not sure how `ARMA_USE_WRAPPER` is getting turned on, but I think if you change `ARMADILLO_INCLUDE_DIR` to be `armadillo-9.800.6/tmp/include/` instead of just `armadillo-9.800.6/include/`, then CMake will detect `ARMA_USE_WRAPPER` as set, and then `ARMADILLO_LIBRARIES` should also include `armadillo.lib`.  Maybe do you want to try that and see what it does?

Reply:
Done